keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,811 @@
+Client-go will support both plugins binaries which are themselves the proxy
+process, and additionally plugin binaries which complete, but are responsible
+for somehow starting a proxy process and returning its listening address.
+
+The client will send its request to the proxy plugin process, and the proxy
+will forward the request to the API server, rather than returning the modified
+request to the client to be sent (see alternatives considered).  The downside
+of this approach is that the proxy must duplicate the some of the transport
+functionality of client packages, but allows for more flexibility for proxy
+plugin developers.
+
+#### Extending ExecConfig
+
+This proposal extends the ExecConfig API object to include configuration
+options for the proxy process.
+
+The ExecConfig would be extended to include a local address, which will be
+passed to the exec plugin.  The exec plugin will create a socket and listen on
+this address, which can either be a Unix domain socket or a TCP/IP socket.
+
+```go
+type ExecConfig struct {
+	// Command to execute.
+	Command string `json:"command"`
+	// Arguments to pass to the command when executing it.
+	// +optional
+	Args []string `json:"args"`
+	// Env defines additional environment variables to expose to the process. These
+	// are unioned with the host's environment, as well as variables client-go uses
+	// to pass argument to the plugin.
+	// +optional
+	Env []ExecEnvVar `json:"env"`
+
+	// Preferred input version of the ExecInfo. The returned ExecCredentials MUST use
+	// the same encoding version as the input.
+	APIVersion string `json:"apiVersion,omitempty"`
+
+	// ...
+
+	// The configuration for an exec proxy.
+	ExecProxy ExecProxyConfig `json:"execProxy"`
+}
+
+```
+
+The ExecProxyConfig includes the address where the proxy is expected to listen.
+```go
+type ExecProxyConfig struct {
+    // Do we need any proxy configuration defined in the kubeconfig?
+}
+```
+
+Input and output from the executed binary.
+```go
+type ExecProxyRequest struct {
+	metav1.TypeMeta
+
+	// +optional
+	Config ExecProxyConfig `json:"config"`
+
+	// +optional
+	// Whether or not the command is being called from an interactive prompt.
+    // A hardware security module might require user interaction on setup,
+	// so in that case the proxy might want to fail if this is false.
+	Interactive bool `json:"interactive"`
+
+	// +optional
+	// Cluster is the destination Kubernetes cluster.  This includes Server
+    // address, TLSServerName, CertificateAuthorityData and ProxyURL.
+	Cluster *Cluster `json:"cluster"`
+}
+
+type ExecProxyResponse struct {
+	metav1.TypeMeta
+
+	Listener Address `json: "address"`
+
+	// +optional
+	CertificateAuthorityData []byte `json:"certificate-authority-data,omitempty"`
+
+	// +optional
+	ClientCertificate []byte `json:"client-certificate"`
+}
+
+type Address struct {
+	// +optional
+	UnixSocket UnixSocket
+
+	// +optional
+	Port int `json:"port"`
```

Comment:
Why not just a URL?

Reply:
I think that makes sense to me and I updated the KEP to reflect that.  @enj initially suggested the port so he might want to add his thoughts here.